spec.md

```diff
@@ -0,0 +1,1465 @@
+4. If `A plays B:I @abstract` and $B(I) < C(J)$ then `A plays B:J` cannot be non-abstract.
+5. If `A owns B @abstract` and $B < C$ then `A owns C` cannot be non-abstract. 
+6. If `A owns B[] @abstract` and $B < C$ then `A owns C[]` cannot be non-abstract. 
+
+**Case VALUES**
+* `A owns B @values(v1, v2)` postulates if $a : A$ then $`a \in \{v_1, v_2\}`$ , ***requiring*** that 
+  * either $`A : \mathbf{Att}`$, $`A < V`$, $`v_i : V`$, 
+  * or $`A`$ is the component of a struct, see section on struct defs. 
+  
+  **Generalizes** to $n$ values.
+* `A owns B @regex(v1..v2)` postulates if $a : A$ then $`a`$ conforms with regex `<EXPR>`.
+* `A owns B @range(v1..v2)` postulates if $a : A$ then $`a \in [v_1,v_2]`$ (conditions as before).
+* `A value B @values(v1, v2)` postulates if $a : A$ then $`a \in \{v_1, v_2\}`$ , ***requiring*** that: 
+  * either $`A : \mathbf{Att}`$, $`A < V`$, $`v_i : V`$, 
+  * or $`A`$ is the component of a struct, see section on struct defs.
+  
+  **Generalizes** to $n$ values.
+* `A value B @regex(v1..v2)` postulates if $a : A$ then $`a`$ conforms with regex `<EXPR>`.
+* `A value B @range(v1..v2)` postulates if $a : A$ then $`a \in [v_1,v_2]`$ (conditions as before).
+
+**Case DISTINCT**
+* `A owns B[] @distinct` postulates that when $`[b_1, ..., b_n] : [B]`$ then all $`b_i`$ are distinct. 
+* `B relates I[] @distinct` postulates that when $`[x_1, ..., x_n] : [I]`$ then all $`x_i`$ are distinct.
+
+### Triggers
+
+**Case DEP_DEL (CASCADE/INDEPEDENT)**
+* `(relation) B relates I @cascade`: deleting $`a : A`$ with existing $`b :_! B(a:I,...)`$, such that $`b :_! B(...)`$ violates $B$'s cardinality for $I$, triggers deletion of $b$.
+  * **defaults** to **TT** error
+* `(relation) B @cascade`: deleting $`a : A`$ with existing $`b :_! B(a:I,...)`$, such that $`b :_! B(...)`$ violates $B$'s cardinality _for any role_ of $B$, triggers deletion of $b$.
+  * **defaults** to **TT** error
+* `(attribute) B @independent`. When deleting $`a : A`$ with existing $`b :_! B(a:O_B)`$, update the latter to $`b :_! B`$.
+  * **defaults** to: deleting $`a : A`$ with existing $`b :_! B(a:O_B)`$ triggers deletion of $b$.
+
+
+### Value types
+
+**Case PRIMITIVES**
+* `bool`
+  * Terms: `true`, `false`
+* `long` — _Comment: still think this could be named more nicely_
+  * Terms: 64bit integers
+* `datetime`
+  * TODO: list formats
+* `time` — _Comment: relative times_
+* `string` — _Comment: dynamically sized type; but could consider Arrow-like implementation, see e.g. [here](https://pola.rs/posts/polars-string-type/)_
+
+**Case STRUCT**
+```
+struct S:
+  C1 value V1 (@values(<EXPR>)),
+  C2 value V2 (@values(<EXPR>));
+```
+adds
+* _Struct type_ $S : \mathbf{Type}$
+* _Struct components_ $`C_1 : \mathbf{Type}`$, $`C_2 : \mathbf{Type}`$, and identify $S = C_1 \times C_2$
+* _Component value casting_: $C_1 < V_1$, $C_2 < V_2$
+* _Component value terms_: whenever $v : V_i$ then $v : C_i$ if $v$ conforms with `<EXPR>`
```

Comment:
I'm assuming here `<EXPR>` can resolve to `v1, v2` or similar? Might be worth defining the permitted values of `<EXPR>` (and `<LABEL>`) somewhere. Likely requires a distinction between types of expression, e.g. `<REGEX>`, `<RNGEX>`, `<TPLEX>` for regular, range, and tuple expressions respectively.

Reply:
Also later: `<AGG>` etc. As a spec, everything should ideally be defined in some manner before use.